### DISCUSSION TITLE
TODO App example is broken

### DISCUSSION THREAD

**Discussion by pedrokarneiro** [pyscript/pyscript#1483] [General]
### TODO App example  is broken

About the TODO App example (https://pyscript.net/examples/todo.html).

I was trying to recreate the TODO App based on its example ([View code](https://pyscript.net/examples/todo.html)), and got this error instantly:

```
Traceback (most recent call last):
  File "/lib/python3.10/_pyodide/_base.py", line 460, in eval_code
    .run(globals, locals)
  File "/lib/python3.10/_pyodide/_base.py", line 306, in run
    coroutine = eval(self.code, globals, locals)
  File "<exec>", line 8, in <module>
  File "/home/pyodide/pyscript/__init__.py", line 270, in select
    el = el.content
AttributeError: 'NoneType' object has no attribute 'content'
```

**Comment by pedrokarneiro** [comment]
My copy of the 'TODO App' project (feel free to comment /  criticize / add / whatever):

https://9fba15fd-aed0-4ee7-8d1f-5443d9d92b91.pyscriptapps.com/bb3bf821-13ce-48df-b1cb-ec9858b93d65/latest/

**Reply by WebReflection** [reply]
You have CORS issues there, this link is not super useful for us as we can't really debug in there, but to me the problem is clear ... see my other reply, thanks!

**Comment by WebReflection** [comment]
This is an interesting one, because what we show in the *View Code* actually implicitly depends on the HTML content of the demo itself, represented by:

```html
                <section>
                    <div class="text-center w-full mb-8">
                        <h1
                            class="text-3xl font-bold text-gray-800 uppercase tracking-tight"
                        >
                            To Do List
                        </h1>
                    </div>
                    <div>
                        <input
                            id="new-task-content"
                            class="py-input"
                            type="text"
                        />
                        <button
                            id="new-task-btn"
                            class="py-button"
                            type="submit"
                            py-click="add_task()"
                        >
                            Add task
                        </button>
                    </div>
 
                    <py-list id="myList"></py-list>
                    <div
                        id="list-tasks-container"
                        class="flex flex-col-reverse mt-4"
                    ></div>
 
                    <template id="task-template">
                        <section class="task py-li-element">
                            <label for="flex items-center p-2 ">
                                <input class="mr-2" type="checkbox" />
                                <p class="m-0 inline"></p>
                            </label>
                        </section>
                    </template>
                </section>
```

and needed by the Python code:

```python
# define the task template that will be use to render new templates to the page
task_template = Element("task-template").select(".task", from_content=True)
task_list = Element("list-tasks-container")
new_task_content = Element("new-task-content")             
```

I don't know how to best move forward in fixing this *View Code* gotcha because it's clear to me users trying our demo that depend on specific template and other DOM nodes won't ever see these within a `<py-tutor>` *View Code* panel.

/cc @FabioRosado @fpliger @antocuni @JeffersGlass any idea what'd be the way forward?

**Comment by fpliger** [comment]
The issue here is that,[ in your example
](https://9fba15fd-aed0-4ee7-8d1f-5443d9d92b91.pyscriptapps.com/bb3bf821-13ce-48df-b1cb-ec9858b93d65/latest/), the whole HTML elements (in the example) are missing. I've copied your app, added the HTML that was missing (that actually defined the HTML elements on your page) and it seems to work fine. You can see my copy [here](https://7589ccb7-f838-4034-abea-5550f0e8d949.pyscriptapps.com/62120e03-9498-4810-a058-d3459f264bc7/latest/).

@WebReflection You are right in that the `py-tutor` is a little weird. I've added in a rush to remove the wall of text on the examples that was showing the source and put it into a plugin. It should be a bit different tbh and it's probably a good idea to revisit it (it actually would be a good idea to revisit the examples and how we host them in general).

**Reply by WebReflection** [reply]
Idea: add an `html` attribute, among the `modules` one, that points at the id of the element that contains related code so we can show that into the *View Code* panel too as *HTML* needed for the example to run.

**Reply by fpliger** [reply]
Ah, yeah. Good call.

I think that having the `py-tutor` NOT be a Custom Element and be a plugin that allow you to specific what you want to include in the tutor via the config. So, for instance:

```toml
[[plugins.py-tutor]]
include = ['html', 'main.py', 'todo.py']
```

or something along these lines..

**Reply by WebReflection** [reply]
I am bit skeptical about that `html` as is ... there's no room for disambiguation with a folder ... maybe this instead?

```
include = ['#html', 'main.py', 'todo.py']
```

at least we have a selector there which hopefully disambiguate more explicitly?


**edit**
that selector points at the DOM but maybe you meant which attribute to use? so we kill modules attribute with this idea?

**Comment by fpliger** [comment]
Yeah yeah, wasn't a formal proposal but "having the ability to... somehow".

+1 on the idea of specifying selectors as well.

> so we kill modules attribute with this idea?

Good question. I see pros and cons..  I'll open a separate discussion so we can discuss there (as I think there are a few extra nuances too).
